zilliqa/src/pool.rs

```diff
@@ -313,6 +331,35 @@ impl TransactionPool {
             Self::remove_from_gas_index(&mut self.gas_index, existing_txn);
             // Remove the existing transaction from `hash_to_index` if we're about to replace it.
             self.hash_to_index.remove(&existing_txn.hash);
+            // Decrease the count of transactions tracked by this sender
+            self.decrease_counter_for_user(existing_txn.signer);
+            true
+        } else {
+            false
+        };
+
+        // If it's a transaction that will increase the count in the pool
+        if !is_replacement {
+            // Check global counter
+            if self.transactions.len() + 1 > self.config.maximum_global_size as usize {
+                return ValidationFailed(ValidationOutcome::GlobalTransactionCountExceeded);
+            }
+            // Check total number of slots for senders
+            if !self.sender_txn_counter.contains_key(&txn.signer)
+                && self.sender_txn_counter.len() + 1
+                    > self.config.total_slots_for_all_senders as usize
+            {
+                return ValidationFailed(ValidationOutcome::TotalNumberOfSlotsExceeded);
+            }
+            // Check per sender counter
+            let sender_counter = self
+                .sender_txn_counter
+                .get(&txn.signer)
+                .copied()
+                .unwrap_or_default();
+            if sender_counter + 1 > self.config.maximum_txn_count_per_sender {
+                return ValidationFailed(ValidationOutcome::TransactionCountExceededForSender);
+            }
         }
```

Comment:
Can we just add this code to the `else` block above and remove `is_replacement`?